src/System.Private.CoreLib/src/System/Text/Utf8Span.Conversion.cs

```diff
@@ -0,0 +1,321 @@
+        /// </summary>
+        /// <remarks>
+        /// The original <see cref="Utf8Span"/> is left unchanged by this operation.
+        /// </remarks>
+        public Utf8String Normalize(NormalizationForm normalizationForm = NormalizationForm.FormC)
+        {
+            // TODO_UTF8STRING: Reduce allocations in this code path.
+
+            return new Utf8String(this.ToString().Normalize(normalizationForm));
+        }
+
+        /// <summary>
+        /// Converts this <see cref="Utf8Span"/> to the desired Unicode normalization form, writing the
+        /// UTF-8 result to the buffer <paramref name="destination"/>.
+        /// </summary>
+        /// <returns>
+        /// The number of bytes written to <paramref name="destination"/>, or -1 if <paramref name="destination"/>
+        /// is not large enough to hold the result of the normalization operation.
+        /// </returns>
+        /// <remarks>
+        /// The original <see cref="Utf8Span"/> is left unchanged by this operation. Note that the the required
+        /// length of <paramref name="destination"/> may be longer or shorter (in terms of UTF-8 byte count)
+        /// than the input <see cref="Utf8Span"/>.
+        /// </remarks>
+        public int Normalize(Span<byte> destination, NormalizationForm normalizationForm = NormalizationForm.FormC)
+        {
+            // TODO_UTF8STRING: Reduce allocations in this code path.
+
+            ReadOnlySpan<char> normalized = this.ToString().Normalize(normalizationForm);
+            OperationStatus status = Utf8.FromUtf16(normalized, destination, out int _, out int bytesWritten, replaceInvalidSequences: false, isFinalBlock: true);
+
+            Debug.Assert(status == OperationStatus.Done || status == OperationStatus.DestinationTooSmall, "Normalize shouldn't have produced malformed Unicode string.");
+
+            if (status != OperationStatus.Done)
+            {
+                bytesWritten = -1; // "destination too small"
+            }
+
+            return bytesWritten;
+        }
+
+        /// <summary>
+        /// Converts this <see cref="Utf8Span"/> to a <see langword="char[]"/>.
+        /// </summary>
+        public unsafe char[] ToCharArray()
+        {
+            if (IsEmpty)
+            {
+                return Array.Empty<char>();
+            }
+
+            // TODO_UTF8STRING: Since we know the underlying data is immutable, well-formed UTF-8,
+            // we can perform transcoding using an optimized code path that skips all safety checks.
+            // We should also consider skipping the two-pass if possible.
+
+            fixed (byte* pbUtf8 = &DangerousGetMutableReference())
+            {
+                byte* pbUtf8Invalid = Utf8Utility.GetPointerToFirstInvalidByte(pbUtf8, this.Length, out int utf16CodeUnitCountAdjustment, out _);
+                Debug.Assert(pbUtf8Invalid == pbUtf8 + this.Length, "Invalid UTF-8 data seen in buffer.");
```

Comment:
>pbUtf8Invalid == pbUtf8 + this.Length [](start = 29, length = 37)

doesn't this assert need to be a real check?

Reply:
No. The data is assumed to be well-formed. (This is part of the contract of the `Utf8Span` type.) If it is not, then the contents of the returned buffer are undefined.